Fix type stripping error for gRPC client in E2E test

Merseleo opened the pull request:
Closes #308

## What I have made

<!-- write down what changes have been made, what was removed/added/changed -->
<!-- e.g. I added a new button, which does x -->

- I added a second test fixture that provides a gRPC client (with a correct connection depending on the browser the test runs in)
- I added a script to run playwright or node without type stripping, which is enabled per fault since Node 23 and the gendered gRPC code violates the new guidelines for Node 23
- I demonstrated the usage of the gRPC client in a cleanup function in a E2E test and adapted the package.json commands for running E2E tests, so it also works with Node 23

## Checklist

Either tick or cross out the items that do not apply (using \~\~example text\~\~) and give a reason why the item does not apply.

- [x] I have updated the documentation accordingly and commented my code
- [x] I have added tests that prove my fix is effective or that my feature works
  ~~- [ ] unit tests~~
  ~~- [ ] integration tests~~
  - [x] end-to-end tests
- [x] (for reviewer) I have checked the implementation against the requirements

Commit:
feat: create additional grpc client in E2E tests

Commit:
feat: disable type stripping when running Node 23.x.x

Commit:
refactor: use grpc client in project creation E2E test to cleanup created projects after test

Slartibartfass2 commented on scripts/run-node.js:
```diff
@@ -0,0 +1,7 @@
+// this script checks whether node is running in version 23 and if so,
+// type stripping is deactivated, since the generated GRPC client does not work with type stripping.
+
+export const isNode23 = process.version.startsWith("v23");
```
What happens in later versions?

Merseleo replied:
I don't know. At the moment this feature is experimental so can still change. Maybe they delete it in Node 24, maybe not, but as far as we dont know what happens in the future I would only distinguish between Node 23 and Node 22 and earlier versions.

Merseleo replied:
But I they keep it **and** protobuf-ts doesn't get updated to conform to the new type stripping **and** somebody updates to Node 24, we probably need to update the script to parse the version number and check whether it is >= 23 or <23.

Slartibartfass2 replied:
I think that's fine for now. Maybe we should declare the used node version for this project.

Merseleo replied:
But the point is that we don't just use one node version, but several. At most, we could list the node version up to which it works in any case (and perhaps the oldest tested version and in between it should also work). I would about this to the README.md or do you have a better solution?

Slartibartfass2 replied:
You can use the [engines field](https://docs.npmjs.com/cli/v11/configuring-npm/package-json#engines) in `package.json` to define the supported range of versions.

Merseleo replied:
Ah thank you for the hint.

Slartibartfass2 commented on scripts/run-node.js:
Is it necessary to put this into a separate file?

Merseleo replied:
No, that's not necessary, I just wanted to keep it apart, because theoretically this problem could also occur with other tools.

Slartibartfass2 commented on scripts/run-playwright.js:
```diff
@@ -0,0 +1,16 @@
+    });
+}
+
+runPlaywright(process.argv.slice(2));
```
Why do we skip the first two arguments?

Merseleo replied:
Short answer: because the person I copied the script from did this. Long answer: You want to skip the first two arguments as they are not relevant and should not be use as _...args_ in the spawn command, because these two arguments are the path of the node executable and of this script (see [here](https://nodejs.org/api/process.html#processargv) for details).

Slartibartfass2 replied:
Please add this explanation as a comment.

Merseleo replied:
I added it, is it clearer now?

Slartibartfass2 replied:
Now it's clear.

Slartibartfass2 commented on tests/e2e/fixtures/general-fixture.ts:
```diff
@@ -34,3 +36,25 @@ export const test = base.extend<{ forEachTest: void } & TestOptions>({
+ * This gRPC client allows to directly send requests to the corresponding
+ * mock backend before / after (or during) the end-to-end test.
+ */
+export const mockBackendServiceTest = base.extend<
+    { mockBackendService: SnowballRClient } & TestOptions
+>({
+    mockBackendUrl: ["http://localhost:3001", { option: true }],
+    mockBackendService: async ({ mockBackendUrl }, use) => {
+        const transport = new GrpcWebFetchTransport({
+            baseUrl: mockBackendUrl,
+        });
+        const mockBackendService = new SnowballRClient(transport);
+        await use(mockBackendService);
+    },
+});
```
Is there a reason why this is separate from the default test fixture?

Merseleo replied:
Yes there are two reasons:
1) if you dont want to use the `mockBackendService` in your tests, you could also import only the `defaultTest` fixture (or vice versa) and
2) in my opinion that are two different use cases and should not be mixed in one fixture. Otherwise I would add two functionalities to one mixture and would need to explain both in one comment.

Slartibartfass2 replied:
Alright, that's fine for me.

Slartibartfass2 commented on tests/e2e/project-creation.test.ts:
```diff
@@ -8,6 +8,12 @@ test.describe("Creating a new project", () => {
         await page.goto("/");
     });
 
+    test.afterAll(async ({ mockBackendService }) => {
+        mockBackendService.softDeleteProject({ id: "0" });
+        mockBackendService.softDeleteProject({ id: "1" });
+        mockBackendService.softDeleteProject({ id: "2" });
+    });
```
Which projects are deleted here?

Merseleo replied:
The projects that are created in this test group. But you are right, only two projects are created, that should be deleted afterwards.

I do this 1) for demonstration how to use the mockBackendService and 2) to "rollback" the changes I made.

Slartibartfass2 replied:
I wonder if we should do this in each test instead of the `afterAll` hook. Not sure what's the best way to clean this up.

Merseleo replied:
In my opinion we should not do this after each test, as this whole test group tests an entire workflow I want to explicitly test, that if I already created a project, the second project is added below the first project and both exist. So if I would cleanup after each test in this case, it would be something different. In general, there are of course some scenarios where we should cleanup after each test. 

Another solution would be add further calls to the API that allow a cleanup, but this shouldn't be content of this PR.

Slartibartfass2 replied:
That's fine for me.

But so far, we do not test if another project was created before, or am I missing something?

Merseleo replied:
No you're not missing anything, you are right. We dont test it and it wouldn't make a difference if I would not cleanup. But then we come back to 1): I wanted to demonstrate how to use the `mockBackendService` and as a nice add-on we get the cleanup if we will test whether another project was created.

Slartibartfass2 reviewed (changes requested):
Great work! I'm excited to use it for my tests.
I tested it with node 22 and 23.

Could you please add more information to the wiki that explains how to use these fixtures.

Commit:
chore: move node version and type stripping flag constants into the run-playwright.js script

Commit:
refactor: only delete created projects after 'Creating new project' E2E test

Merseleo reviewed (commented):
I hope I could answer all your questions.

Slartibartfass2 reviewed (changes requested):
Some minor things left.

Slartibartfass2 commented:
Don't forget to rebase when you're finished.

Commit:
docs: extend comments in 'run-playwright.js' script

Add comment for the runPlayright method to explain what this functions does and explain why the two first arguments from argv are skipped

Slartibartfass2 reviewed (changes requested):
Nearly done!

Commit:
chore: add devEngines property to package.json

Define the required node version for development of the frontend; set to >=22 <=23.11

Merseleo reviewed (commented):
Thank you for the hints.

Commit:
docs: extend section about E2E tests in the frontend wiki

Slartibartfass2 commented on wiki/Contributing.md:
```diff
@@ -103,21 +104,65 @@ instead of classes or text content.
+
+#### Requirements
+
+The E2E tests require a mock-backend to be running for each supported browser (Chromium, Firefox, and WebKit).
+These mock-backends simulate the backend responses.
+
+To manually start all necessary mock-backends, you can use the provided script, which will start one instance per
+browser if you provide the three ports:
```
What was that again? "Keep it consistent"? :D
```suggestion
The E2E tests require a mock backend to be running for each supported browser (Chromium, Firefox, and WebKit).
These mock backends simulate the backend responses.

To manually start all necessary mock backends, you can use the provided script, which will start one instance per
browser if you provide the three ports:
```

Merseleo replied:
You are completely right, I fixed it.

Slartibartfass2 commented on wiki/Contributing.md:
```diff
@@ -103,21 +104,65 @@ instead of classes or text content.
 
+#### Best Practices
+
+1. **Use the test fixture**:
+   Instead of importing the `test` fixture directly from "@playwright/test" directly,
+   use the `test` variable exported from `fixtures/general-fixture.ts`.
+   This ensures that the API calls to the mock-backend are automatically redirected to the correct mock-backend
+   based on the browser the test is running in.
+
+   In general, it makes sense to create an own fixture, if you want to use an object in multiple tests or
+   run a certain function before / after each test.
+
+2. **Use Page Object Models (POMs)**:
+   To make tests more maintainable and readable, use page object models (pom) to encapsulate UI logic.
+   For example, the `pom/create-project-dialog-model.ts` file wraps a page with custom selectors and
+   helper functions for the `CreateProjectDialog` component,
+   improving clarity and reusability across tests. In particular, these _pom_s can be wrapped in a fixture
+   so that they can be accessed directly in each test without having to create a separate \_pom_ in each test.
```
Let's link to the files so that they can be viewed in the browser.
```suggestion
1. **Use the test fixture**:
   Instead of importing the `test` fixture directly from `@playwright/test` directly,
   use the `test` variable exported from
   [`fixtures/general-fixture.ts`](https://github.com/SE-UUlm/snowballr-frontend/blob/6ac47c3f0330d72d63fe977e476e51ba29937c7b/tests/e2e/fixtures/general-fixture.ts).
   This ensures that the API calls to the mock backend are automatically redirected to the correct mock backend
   based on the browser the test is running in.

   In general, it makes sense to create an own fixture, if you want to use an object in multiple tests or
   run a certain function before / after each test.

2. **Use Page Object Models (POMs)**:
   To make tests more maintainable and readable, use page object models (POMs) to encapsulate UI logic.
   For example, the [`pom/create-project-dialog-model.ts`](https://github.com/SE-UUlm/snowballr-frontend/blob/6ac47c3f0330d72d63fe977e476e51ba29937c7b/tests/e2e/pom/create-project-dialog-model.ts)
   file wraps a page with custom selectors and helper functions for the `CreateProjectDialog` component,
   improving clarity and reusability across tests. In particular, these POMs can be wrapped in a fixture
   so that they can be accessed directly in each test without having to create a separate POM in each test.
```

Merseleo replied:
I agree but I wouldn't add this link with the sha, but the branch name develop as these links would work always after the merge, are you ok with that?

Slartibartfass2 reviewed (dismissed):
Only minor things

Commit:
docs: fix typos and add links to example fixture and pom files in Contributing.md
